File: ascii/draw_image.py
import sys
import os
from PIL import Image
import numpy as np

# define the set of characters to use
chars = ' .:-=+*#%@$'

# get an image path
path = 'cat.jpg'
if len(sys.argv) > 1:
	path = sys.argv[1]

# get terminal size
rows, columns = os.popen('stty size', 'r').read().split()
rows, columns = int(rows), int(columns)

# load the image convert to grayscale
img = np.array(Image.open(path))
img = img.mean(axis=2)
img /= img.max()

scale = (img.shape[0] / rows, img.shape[1] / columns)

# print the image as ASCII characters
result = ''
for x in range(rows):
	for y in range(columns):
		index = (int(x * scale[0]), int(y * scale[1]))
		value = np.mean(img[index[0], index[1]])
		result += chars[int(value * len(chars)) - 1]
	result += '\n'
print(result, end='')

File: ascii/draw_image.cpp
#include <string>
#include <iostream>
#include <sys/ioctl.h>
#include <CImg.h>

using namespace std;
using namespace cimg_library;

int main(int argc, char** argv) {
	// define the set of characters to use
	const string chars = " .:-=+*#%@$";

	// get the image path
	string path = "img.jpg";
	if(argc > 1) {
		path = argv[1];
	}

	// load the image
	CImg<unsigned char> img(path.c_str());

	// get the terminal dimensions
    struct winsize dims;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &dims);

	// print the image as ASCII characters
	float xScale = img.width() / dims.ws_col;
	float yScale = img.height() / dims.ws_row;
	for(int y = 0; y < dims.ws_row; y++) {
		for(int x = 0; x < dims.ws_col; x++) {
			int index = y * yScale * img.width() + x * xScale;
			float value = img[index] / 255.0;
			cout << chars[(int) (value * chars.length())];
		}
		cout << "\n";
	}
	return 0;
}
